dagql/idtui/frontend.go

```diff
@@ -340,41 +375,47 @@ func (r renderer) renderDuration(out *termenv.Output, span *Span) {
 // 	return nil
 // }
 
-type spanFilter struct {
-	db               *DB
-	gcThreshold      time.Duration
-	tooFastThreshold time.Duration
-}
+const (
+	ShowCompletedVerbosity    = 1
+	ShowInternalVerbosity     = 2
+	ShowEncapsulatedVerbosity = 2
+	ShowSpammyVerbosity       = 3
+	ShowDigestsVerbosity      = 3
 
-func (sf spanFilter) shouldShow(opts FrontendOpts, row *TraceRow) bool {
+	DefaultVerbosity = ShowCompletedVerbosity
+)
+
+func (opts FrontendOpts) ShouldShow(tree *TraceTree) bool {
 	if opts.Debug {
 		// debug reveals all
 		return true
 	}
-	span := row.Span
-	if span.IsInternal() && opts.Verbosity < 2 {
+	span := tree.Span
+	if span.IsInternal() && opts.Verbosity < ShowInternalVerbosity {
 		// internal steps are hidden by default
 		return false
 	}
-	if row.Parent != nil && (span.Encapsulated || row.Parent.Span.Encapsulate) && row.Parent.Span.Err() == nil && opts.Verbosity < 2 {
+	if tree.Parent != nil && (span.Encapsulated || tree.Parent.Span.Encapsulate) && tree.Parent.Span.Err() == nil && opts.Verbosity < ShowEncapsulatedVerbosity {
 		// encapsulated steps are hidden (even on error) unless their parent errors
 		return false
 	}
 	if span.Err() != nil {
 		// show errors
 		return true
 	}
-	if sf.tooFastThreshold > 0 && span.Duration() < sf.tooFastThreshold && opts.Verbosity < 3 {
-		// ignore fast leaf steps; signal:noise is too poor
-		if row.Parent != nil && row.Parent.Span.SpanContext().SpanID() != sf.db.PrimarySpan {
-			return false
-		}
+	if opts.TooFastThreshold > 0 && span.Duration() < opts.TooFastThreshold && opts.Verbosity < ShowSpammyVerbosity {
+		// TODO(vito): bring this back
+		// if tree.Parent != nil && tree.Parent.Span.SpanContext().SpanID() != opts.PrimarySpan {
+		// 	return false
+		// }
+		// ignore fast steps; signal:noise is too poor
+		return false
```

Comment:
Are we intentionally not bringing this back before merging? Ideally I'd still like it, it made quite a few steps just never appear if they were quick enough.

Reply:
Didn't mean to leave this commented (it just needed access to more info post-refactor), but I do find the heuristic a little confusing and don't know how it should be updated for the new UI. Why do we want to show really fast spans beneath the primary span in particular, as opposed to the usual "really fast = boring/noise" logic? Does this affect the Plain UI more than the regular TUI? Does it affect the primary span in a way that's unique to the primary span and not other calls? Do we want this to apply to zoomed spans, too?

I'm trying to tease apart whether this is really a heuristic that belongs here or if it's working around a gap in other parts of the UI, like maybe when you select a span we can have a key to show its full recipe in a separate pane or something.

Reply:
Brought this back in the last commit to keep the PR unblocked but yeah, still find it a bit confusing. Found myself wondering if other things should also be revealed in this scenario (like internal steps) and couldn't intuit a rhyme or reason.

Reply:
Yeah, this is tricky, the heuristic definitely isn't perfect but there is a reason for it.

Suppose I have `dagger call do-thing process-it export --path=thing`. Ideally, my top-level spans should look *something* like:

- `MyModule`
- `MyModule.DoThing`
- `Thing.ProcessIt`
- `Directory.Export`

Each of these is tied directly to a CLI subcommand, so they should *always* show. The reason I originally noticed this was because the `export` was sometimes showing, sometimes not, and was essentially racy depending on the size of the file.

I think we should always show these user subcommand spans at the top level, which translated into the heuristic of always showing the top-level.